File: rust_nom/src/lexer/lexer.rs
use anyhow::Result;
use nom::{
    branch::alt,
    character::complete::{alpha1, char, digit1, multispace0},
    combinator::map,
    IResult,
};

#[derive(Debug, PartialEq)]
pub enum Token {
    Ident(String),
    Int(String),

    Illegal,
    Eof,
    Equal,
    Plus,
    Comma,
    Semicolon,
    LParen,
    RParen,
    LSquirly,
    RSquirly,
    Function,
    Let,
}

pub struct Lexer<'a> {
    input: &'a str,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Lexer<'a> {
        Lexer { input }
    }

    pub fn next_token(&mut self) -> Result<Token, String> {
        let (input, _) = multispace0::<&str, nom::error::Error<&str>>(self.input).unwrap();
        self.input = input;

        if self.input.is_empty() {
            return Ok(Token::Eof);
        }

        let (input, token) = alt((Self::one_char_token, Self::ident, Self::integer))(self.input)
            .map_err(|_| "Invalid token".to_string())?;

        self.input = input;

        Ok(token)
    }

    fn one_char_token(input: &str) -> IResult<&str, Token> {
        alt((
            map(char('{'), |_| Token::LSquirly),
            map(char('}'), |_| Token::RSquirly),
            map(char('('), |_| Token::LParen),
            map(char(')'), |_| Token::RParen),
            map(char(','), |_| Token::Comma),
            map(char(';'), |_| Token::Semicolon),
            map(char('+'), |_| Token::Plus),
            map(char('='), |_| Token::Equal),
        ))(input)
    }

    fn ident(input: &str) -> IResult<&str, Token> {
        map(alpha1, |s: &str| match s {
            "fn" => Token::Function,
            "let" => Token::Let,
            _ => Token::Ident(s.into()),
        })(input)
    }

    fn integer(input: &str) -> IResult<&str, Token> {
        map(digit1, |s: &str| Token::Int(s.into()))(input)
    }
}

#[cfg(test)]
mod test {
    use super::{Lexer, Token};
    use anyhow::Result;

    #[test]
    fn get_next_token() -> Result<()> {
        let input = "=+(){},;";
        let mut lexer = Lexer::new(input);

        let tokens = vec![
            Token::Equal,
            Token::Plus,
            Token::LParen,
            Token::RParen,
            Token::LSquirly,
            Token::RSquirly,
            Token::Comma,
            Token::Semicolon,
        ];

        for token in tokens {
            let next_token = lexer.next_token().unwrap();
            println!("expected: {:?}, received {:?}", token, next_token);
            assert_eq!(token, next_token);
        }

        return Ok(());
    }

    #[test]
    fn get_next_complete() -> Result<()> {
        let input = r#"let five = 5;
            let ten = 10;
            let add = fn(x, y) {
                x + y;
            };
            let result = add(five, ten);"#;

        let mut lex = Lexer::new(input);

        let tokens = vec![
            Token::Let,
            Token::Ident(String::from("five")),
            Token::Equal,
            Token::Int(String::from("5")),
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("ten")),
            Token::Equal,
            Token::Int(String::from("10")),
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("add")),
            Token::Equal,
            Token::Function,
            Token::LParen,
            Token::Ident(String::from("x")),
            Token::Comma,
            Token::Ident(String::from("y")),
            Token::RParen,
            Token::LSquirly,
            Token::Ident(String::from("x")),
            Token::Plus,
            Token::Ident(String::from("y")),
            Token::Semicolon,
            Token::RSquirly,
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("result")),
            Token::Equal,
            Token::Ident(String::from("add")),
            Token::LParen,
            Token::Ident(String::from("five")),
            Token::Comma,
            Token::Ident(String::from("ten")),
            Token::RParen,
            Token::Semicolon,
            Token::Eof,
        ];

        for token in tokens {
            let next_token = lex.next_token().unwrap();
            println!("expected: {:?}, received {:?}", token, next_token);
            assert_eq!(token, next_token);
        }

        return Ok(());
    }
}

File: typescript/lexer/__tests__/lexer.ts
import { TokenType, Tokenizer } from "..";

test("test getNextToken()", function() {
    const input = `=+(){},;`;

    const tokens = [
        TokenType.Equal,
        TokenType.Plus,
        TokenType.LParen,
        TokenType.RParen,
        TokenType.LSquirly,
        TokenType.RSquirly,
        TokenType.Comma,
        TokenType.Semicolon,
    ];

    const lexer = new Tokenizer(input);

    for (const token of tokens) {
        expect(lexer.getNextToken().type).toBe(token);
    }

});


test("test getNextToken() complete", function() {
    const input =`let five = 5;
        let ten = 10;
        let add = fn(x, y) {
            x + y;
        };
        let result = add(five, ten);`;

    var lex = new Tokenizer(input);

    var tokens = [
        {type: TokenType.Let, literal: "let"},
        { type: TokenType.Ident, literal: "five" },
        {type: TokenType.Equal, literal: "="},
        { type: TokenType.Int, literal: "5" },
        {type: TokenType.Semicolon, literal: ";"},
        {type: TokenType.Let, literal: "let"},
        { type: TokenType.Ident, literal: "ten" },
        {type: TokenType.Equal, literal: "="},
        { type: TokenType.Int, literal: "10" },
        {type: TokenType.Semicolon, literal: ";"},
        {type: TokenType.Let, literal: "let"},
        { type: TokenType.Ident, literal: "add" },
        {type: TokenType.Equal, literal: "="},
        {type: TokenType.Function, literal: "fn"},
        {type: TokenType.LParen, literal: "("},
        { type: TokenType.Ident, literal: "x" },
        {type: TokenType.Comma, literal: ","},
        { type: TokenType.Ident, literal: "y" },
        {type: TokenType.RParen, literal: ")"},
        {type: TokenType.LSquirly, literal: "{"},
        { type: TokenType.Ident, literal: "x" },
        {type: TokenType.Plus, literal: "+"},
        { type: TokenType.Ident, literal: "y" },
        {type: TokenType.Semicolon, literal: ";"},
        {type: TokenType.RSquirly, literal: "}"},
        {type: TokenType.Semicolon, literal: ";"},
        {type: TokenType.Let, literal: "let"},
        { type: TokenType.Ident, literal: "result" },
        {type: TokenType.Equal, literal: "="},
        { type: TokenType.Ident, literal: "add" },
        {type: TokenType.LParen, literal: "("},
        { type: TokenType.Ident, literal: "five" },
        {type: TokenType.Comma, literal: ","},
        { type: TokenType.Ident, literal: "ten" },
        {type: TokenType.RParen, literal: ")"},
        {type: TokenType.Semicolon, literal: ";"},
        {type: TokenType.Eof, literal: "eof"},
    ];

    for (const token of tokens) {
        expect(lex.getNextToken()).toEqual(token);
    }
});
